server/src/main/java/org/candlepin/model/PoolCurator.java

```diff
@@ -2006,4 +2008,115 @@ public void clearPoolSourceEntitlementRefs(Iterable<String> poolIds) {
 
         return output;
     }
+
+    /**
+     * Fetches the product content of all products (including provided & derived provided
+     * products) associated with active pool.
+     *
+     * @param ownerId
+     *  Owner Id
+     *
+     * @return
+     *  A Set of product content associated with active pool.
+     */
+    public Set<ProductContent> getProductContentOfActivePoolsByOwnerId(String ownerId) {
+        Set<ProductContent> productsContent = new HashSet<>();
+        Date currentDate = new Date();
+
+        long activePoolCount = this.getActivePoolCountByOwnerId(ownerId);
+        int blockSize = getBatchBlockSize();
+        int blockCount = (int) Math.ceil(activePoolCount / (float) blockSize);
+
+        CriteriaBuilder criteriaBuilder = this.getEntityManager().getCriteriaBuilder();
+        CriteriaQuery<Pool> query = criteriaBuilder.createQuery(Pool.class);
+        Root<Pool> pool = query.from(Pool.class);
+        query.select(pool);
+        List<Predicate> predicates = new ArrayList<>();
+        predicates.add(criteriaBuilder.equal(pool.get("owner").get("id"), ownerId));
+        predicates.add(criteriaBuilder.greaterThanOrEqualTo(pool.get("endDate"), currentDate));
+        predicates.add(criteriaBuilder.lessThanOrEqualTo(pool.get("startDate"), currentDate));
+        Predicate[] predicateArray = new Predicate[predicates.size()];
+        query.where(predicates.toArray(predicateArray));
+        List<javax.persistence.criteria.Order> orderList = new ArrayList();
+        orderList.add(criteriaBuilder.desc(pool.get("startDate")));
+        query.orderBy(orderList);
+        int offset = 0;
+
+        if (blockCount > 1) {
+            while (blockCount > 0) {
```

Comment:
The blocking here probably isn't critical in our general case, but it also doesn't hurt anything.

However, the code goes out of its way to have special handling for the 1 block case, when that's unnecessary -- even with only a single block, setting the max results and first result in the query won't change the performance of the query enough to justify having separate branches for each case.

Additionally, the block count bit is largely unnecessary. Once you have the element count, block size, and keep track of the offset, the counting of blocks becomes extraneous -- just check that your offset is lower than your element count. Even without the element count, you can achieve the same level of performance by just querying until you get a batch that has zero elements.

That said, fetching the element count is basically the same query here except without returning the pool information -- the DB work is effectively the same. Moreover, the collection processing below runs into not one, but two n+1 problems of doing a ton of extra queries to fetch the product and content collections. I would bet this is overall slower than just fetching all of the "active" pool IDs, then using those to get the related product UUIDs, and then using that to get the content objects. You'd get the whole thing done in three queries, rather than the 2n^p+b+1 queries we have here.

Reply:
Yes it is slower, firing extra queries under the hood to fetch product & contents, this is what I was afraid of hence I chose to execute queries to get Product UUIDs.

> If the goal here is to fetch products for active pools, we should do exactly that: fetch active pools and then pull the products from them in code.

From your previous comment I thought you meant to get the products (and content) of active pool right from the code. 
Maybe we can go back to previous approach of having direct queries to fetch all associated products UUIDs of active pools and then fetch the product's content. What are your thoughts on that ? Also do we care if fetch these UUIDs in batch or not ?

Reply:
Sorry, Github isn't showing me the old code, so I'm going from memory here...

Anyway, IIRC, the old code did pool iteration as a one-off in the method itself, so my assumption was that it would be augmented to simply pull the objects off it there and we'd drop the custom SQL builder code derived from a deprecated method that had a very dangerous oversight in it. My overall stance is that if we're going to add a method to do this operation at the curator level, it's something that should be generalized and as efficient as reasonably possible since it will be available to the rest of the code base at that point, and we might attempt to use it elsewhere in the future.

Once we are working at the curator level, the three queries I mentioned above can be used to fetch the information quicker than object iteration, and with minimal fetching of DB data. One of them will require raw SQL, but we should avoid constructing SQL statements via string concatenation without careful consideration of all the problems that introduces.

Regarding the batching question, there are two cases we want to do requests in batches:
1) if the number of input parameters would exceed the limits of either an IN operator or the overall query parameter limit (this is the most common scenario)
2) optimizing lookups when fetching lazy collections off of certain entities (rare thus far in Candlepin)

I believe in this case, once we move the operation to the curator level, we will need to batch each query due to the nature of the operation (N pools, M products, O content). The easiest way to do this is with the partition method provided by AbstractHibernateCurator, or you can use the partitioning functionality directly. There are other ways to batch, but I would recommend them only in specific cases where your data store(s) or operations don't lend themselves nicely to the partitioning functionality.

Hope that helps, and my apologies for seemingly going back and forth here.

Reply:
Changes made to cut back the complex batching logic & to simplify the SQL queries with JPA criteria queries. I broke the steps into multiple parts of pulling the active PoolIDs, then product & derived product UUIDs, provided product UUIDs and finally derived provided product UUIDs. In between these queries we process for product content so that we can avoid holding everything in-memory.